src/daemon/ca/publishing.rs

```diff
@@ -1,232 +1,1223 @@
+            _ => return Err(Error::publishing("published resource class in the wrong key state")),
+        };
+
+        let staging_set = BasicKeyObjectSet::create(key, timing, signer)?;
+
+        self.keys = ResourceClassKeyState::Staging(StagingKeyState {
+            staging_set,
+            current_set,
+        });
+
+        Ok(())
     }
 
-    pub fn added_or_updated(&self) -> AddedOrUpdated {
-        let name = self.name.clone();
-        let object = self.current.clone();
-        match self.old.clone() {
-            None => AddedOrUpdated::Added(AddedObject::new(name, object)),
-            Some(old) => AddedOrUpdated::Updated(UpdatedObject::new(name, object, old)),
+    fn keyroll_activate(&mut self, timing: &IssuanceTimingConfig, signer: &KrillSigner) -> KrillResult<()> {
+        self.keys = match &self.keys {
+            ResourceClassKeyState::Staging(state) => {
+                let old_set = state.current_set.retire(timing, signer)?;
+                let current_set = state.staging_set.clone().into();
+                ResourceClassKeyState::Old(OldKeyState { current_set, old_set })
+            }
+            _ => return Err(Error::publishing("published resource class in the wrong key state")),
+        };
+
+        Ok(())
+    }
+
+    fn keyroll_finish(&mut self) -> KrillResult<Option<RepositoryContact>> {
+        match self.keys.clone() {
+            ResourceClassKeyState::Old(old) => {
+                let current_set = old.current_set;
+                self.keys = ResourceClassKeyState::current(current_set);
+                Ok(old.old_set.old_repo)
+            }
+            _ => Err(Error::publishing("published resource class in the wrong key state")),
+        }
+    }
+
+    fn update_received_cert(&mut self, updated_cert: &RcvdCert) -> KrillResult<()> {
+        self.keys.update_received_cert(updated_cert)
+    }
+
+    fn update_roas(
+        &mut self,
+        roa_updates: &RoaUpdates,
+        timing: &IssuanceTimingConfig,
+        signer: &KrillSigner,
+    ) -> KrillResult<()> {
+        match self.keys.borrow_mut() {
+            ResourceClassKeyState::Current(state) => state.current_set.update_roas(roa_updates, timing, signer),
+            ResourceClassKeyState::Staging(state) => state.current_set.update_roas(roa_updates, timing, signer),
+            ResourceClassKeyState::Old(state) => state.current_set.update_roas(roa_updates, timing, signer),
+        }
+    }
+
+    fn update_certs(
+        &mut self,
+        cert_updates: &ChildCertificateUpdates,
+        timing: &IssuanceTimingConfig,
+        signer: &KrillSigner,
+    ) -> KrillResult<()> {
+        match self.keys.borrow_mut() {
+            ResourceClassKeyState::Current(state) => state.current_set.update_certs(cert_updates, timing, signer),
+            ResourceClassKeyState::Staging(state) => state.current_set.update_certs(cert_updates, timing, signer),
```

Comment:
I'm not sure what the logic should be here but is it correct to update the "current" set in all cases? In the staging case should the "staging" set be updated?